hal.zig

```diff
@@ -8,11 +8,91 @@ pub const Pin = enum(u8) { // or whatever fits for the HAL
 
     /// Runtime lookup of a pin by name.
     pub fn resolve(name: []const u8) ?Pin;
+
+    /// The corresponding UART instance this pin can be configured to.
+    /// - However... What happens when multiple Uarts can be routed to the same pin?
+    /// - There's an argument this function shouldn't exist, and it's on the user
+    ///   to correctly configure the corresponding pins before using the UART driver,
+    ///   and specify the correct Uart instance themselves.
+    pub fn to_uart_instance(self: Pin) !uart.Instance;
 };
 
-pub const UART = struct {
-    pub fn write_blockingly(uart: UART, chunk: []const u8) !usize;
-    pub fn read_blockingly(uart: UART, chunk: []const u8) !usize;
+// Pretending we're in uart.zig
+pub const uart = struct {
+
+    // Seems like a useful abstraction instead of limiting via a "uN" integer type,
+    // there's usually a very reasonable number of UARTs on any one chip so this
+    // enum shouldn't be too tedious to populate by hand.
+    pub const Instance = enum {
+        uart0,
+        uart1,
+        uart2,
+    };
```

Comment:
How could we handle different properties of different uarts like "uart0 and uart2 are of type a, but uart1 is completely different" (LPC1768, or even worse: AVR)

Reply:
Hmmm little confused what you mean, like the UART peripheral itself varies by instance (USART vs UART on ST for instance), or you're talking to different devices on each instance?